Gems/ROS2/Code/Source/Manipulator/MotorizedJointComponent.cpp

```diff
@@ -0,0 +1,273 @@
+                AZ::TransformBus::Event(
+                    m_debugDrawEntity, &AZ::TransformBus::Events::SetLocalTM, transform * m_debugDrawEntityInitialTransform);
+            }
+            else
+            {
+                AZ_Assert(false, "Not implemented");
+            }
+        }
+
+        const uint64_t deltaTimeNs = deltaTime * 1'000'000'000;
+        float speed_control = m_pidPos.ComputeCommand(control_position_error, deltaTimeNs);
+
+        if (measurement <= m_limits.first)
+        {
+            // allow only positive control
+            speed_control = AZStd::max(0.f, speed_control);
+        }
+        else if (measurement >= m_limits.second)
+        {
+            // allow only negative control
+            speed_control = AZStd::min(0.f, speed_control);
+        }
+
+        if (m_debugPrint)
+        {
+            AZ_Printf(
+                "MotorizedJointComponent",
+                " %s | pos: %f | err: %f | cntrl : %f | set : %f |\n",
+                GetEntity()->GetName().c_str(),
+                measurement,
+                control_position_error,
+                speed_control,
+                m_setpoint);
+        }
+        SetVelocity(speed_control, deltaTime);
+    }
+
+    float MotorizedJointComponent::ComputeMeasurement(AZ::ScriptTimePoint time)
+    {
+        AZ::Transform transform;
+        if (!m_measurementReferenceEntity.IsValid())
+        {
+            AZ::TransformBus::EventResult(transform, this->GetEntityId(), &AZ::TransformBus::Events::GetLocalTM);
+        }
+        else
+        {
+            AZ::Transform transformStepParent;
+            AZ::TransformBus::EventResult(transformStepParent, m_measurementReferenceEntity, &AZ::TransformBus::Events::GetWorldTM);
+            AZ::Transform transformStepChild;
+            AZ::TransformBus::EventResult(transformStepChild, GetEntityId(), &AZ::TransformBus::Events::GetWorldTM);
+            transform = transformStepParent.GetInverse() * transformStepChild;
+        }
+        if (m_linear)
+        {
+            const float last_position = m_currentPosition;
+            m_currentPosition = transform.GetTranslation().Dot(this->m_jointDir);
+            if (m_lastMeasurementTime > 0)
+            {
+                double delta_time = time.GetSeconds() - m_lastMeasurementTime;
+                m_currentVelocity = (m_currentPosition - last_position) / delta_time;
+            }
+            m_lastMeasurementTime = time.GetSeconds();
+            return m_currentPosition;
+        }
+        AZ_Assert(false, "it is not implemented");
+        return 0;
+    }
+
+    void MotorizedJointComponent::SetVelocity(float velocity, float deltaTime)
+    {
+        if (m_animationMode)
+        {
+            ApplyLinVelAnimation(velocity, deltaTime);
+        }
+        else
+        {
+            deltaTime = AZStd::min(deltaTime, 0.1f); // limit max force for small FPS
```

Comment:
Is this comment correct with regards to this action? (not clear)

Reply:
Done